### PR TITLE
Fix memory context usage race condition in partitioning spiller

### PR DESCRIPTION
**Pull request by findepi**

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/spiller/TestGenericPartitioningSpiller.java [hunk 1]
```diff
@@ -139,6 +132,17 @@ private void assertSpilledPages(
...
     }
 
+    private static AggregatedMemoryContext mockMemoryContext()
+    {
+        return new AggregatedMemoryContext();
+    }
+
+    private static SpillContext mockSpillContext()
+    {
+        return bytes -> {
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/spiller/TestGenericPartitioningSpiller.java:222]: what does `{}` mean here? I know that this method is extracted from code which was here already, but still I do not understand this lambda.

File: presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java [hunk 2]
```diff
@@ -178,7 +178,7 @@ private synchronized SingleStreamSpiller getSpiller(int partition)
                      * of memory allocated for pages. We can remove or leave--then we need to make it @ThreadSafe or provide synchronized wrapper (so that thread safety cost
                      * is paid only here).
                      */
-                    spillerFactory.create(types, spillContext.newLocalSpillContext(), memoryContext.newLocalMemoryContext())
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java]: o_O 

I think we need to refactor this as we discussed offline

File: presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
+/*
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java:1]: I would squash that with the commit that is using this. It would be easier to review this class with the usage context.

File: presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    /**
+     * Creates synchronized wrapper for the {@code delegate} context. The delegate won't be closed when returned instance is closed.
+     */
+    public static SynchronizedAggregatedMemoryContext synchronizedMemoryContext(AbstractAggregatedMemoryContext delegate)
+    {
+        return new SynchronizedAggregatedMemoryContext(delegate);
+    }
+
+    private final AbstractAggregatedMemoryContext delegate;
+    private final Closer closer = Closer.create();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java:38]: I would have two inner classes one with closer and without

File: presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+    private SynchronizedAggregatedMemoryContext(AbstractAggregatedMemoryContext delegate)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java]: put this constructor one method below

File: presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.io.IOException;
+
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public final class SynchronizedAggregatedMemoryContext
+        extends AbstractAggregatedMemoryContext
+        implements AutoCloseable
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java:27]: I think you mixed two responsibilities here:
 - synchronize wrapper
 - ownership aware factory

File: presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java [hunk 7]
```diff
@@ -139,13 +145,25 @@ private synchronized IntArrayList partitionPage(Page page, IntPredicate spillPar
...
     }
 
-    private synchronized ListenableFuture<?> flush()
+    private ListenableFuture<?> flushFullBuilders()
+    {
+        return flush(PageBuilder::isFull);
+    }
+
+    @VisibleForTesting
+    ListenableFuture<?> flush()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java:154]: flushUnconditionally?

File: presto-main/src/test/java/com/facebook/presto/spiller/TestGenericPartitioningSpiller.java [hunk 8]
```diff
@@ -57,7 +59,6 @@
     private static final int FOURTH_PARTITION_START = 20;
 
     private static final List<Type> TYPES = ImmutableList.of(BIGINT, VARCHAR, DOUBLE, BIGINT);
-    private static final PartitionFunction POSITIVE_AND_NEGATIVE_PARTITIONS_HASH_GENERATOR = new FourPartitionsPartitionFunction(0);
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/spiller/TestGenericPartitioningSpiller.java:50]: is this related?

**Review by kokosing — APPROVED** [review]

File: presto-main/src/test/java/com/facebook/presto/spiller/TestGenericPartitioningSpiller.java (hunk 1 above)
Comment (findepi) [reply presto-main/src/test/java/com/facebook/presto/spiller/TestGenericPartitioningSpiller.java:222]: It's funny (over-)use of the fact `SpillContext` has single method `updateBytes`, so `{}` means "ignore".

File: presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java (hunk 2 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java]: I made a separate PR with that (#8482). I don't want to bind urgent fixes needed for our release with that refactoring.

File: presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java (hunk 3 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java:1]: As you wish, but mind that the fixes are fixups and will end up as part of the commit adding partitioning spiller.

File: presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java (hunk 4 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java:38]: That would be cleaner but way more verbose. I don't feel convinced.

File: presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java (hunk 6 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java:27]: See my answer below, i.e.

> That would be cleaner but way more verbose. I don't feel convinced.

File: presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java (hunk 7 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java:154]: Most of the times `flush` is just flush, unconditionally.

File: presto-main/src/test/java/com/facebook/presto/spiller/TestGenericPartitioningSpiller.java (hunk 8 above)
Comment (findepi) [reply presto-main/src/test/java/com/facebook/presto/spiller/TestGenericPartitioningSpiller.java:50]: Hardly related to the PR. I refactored the test class while adding more tests. This is all fixups.
